index.bs

```diff
@@ -318,7 +233,190 @@ For example,
 is more readable than
 <code highlight="js">window.scrollBy(50, 0)</code>.
 
-<h3 id="promises">Design asynchronous APIs using Promises</h3>
+<h2 id="api-choices">Designing idiomatic and useful APIs</h2>
+
+<h3 id="constructors">Classes should have constructors when possible</h3>
+
+By default, [[!WEBIDL]] interfaces are reified in JavaScript as "non-constructible" classes: trying
+to create instances of them using the normal pattern in JavaScript, `new X()`, will throw a
+{{TypeError}}. This is a poor experience for JavaScript developers, and doesn't match the design of
+most of the JavaScript standard library or of the classes JavaScript developers themselves create.
+
+From a naive perspective, it's not even coherent: how can instances of this thing ever exist, if
+they can't be constructed? The answer is that the platform has magic powers: it can create instances
+of Web IDL-derived classes without ever actually calling their constructor.
+
+You should strive to reduce such magic in designing your own APIs, and give them constructors which
+allow JavaScript developers to create them, just like the platform does. This means adding an
+appropriate [{{Constructor}}] extended attribute to your interface, and defining the algorithm for
+creating new instances of your class.
+
+Apart from reducing the magic in the platform, adding constructors allows JavaScript developers to
+create instances of the class for their own purposes, such as testing, mocking, or interfacing with
+third-party libraries which accept instances of that class. Additionally, because of how
+JavaScript's class design works, it is only possible to create a subclass of the class if it is
+constructible.
```

Comment:
Because a class can be so easily defined in IDL, and because those classes end up in all global scopes, there needs to be strong advice to really generalize classes for maximum reuse (i.e., WG should talk to the community!)... and that "data only" classes, should just use Dictionaries (i.e., objects). Payment Request has the worst of both, for example: `PaymentAddress` is non-constructible, and pure data... it's tragic :(

Reply:
You should still be able to fix that, no?

Reply:
If you mean adding a Constructor on these interfaces, then yes. We should probably do that -as it would not break anything, I think.

Reply:
I meant turning them into a dictionary if they're just data.

Reply:
No, unfortunately. Couple of reasons:

1. It would be a significant breaking change - as, IIRC, IDL attributes can't return Dictionaries, the API would need some sort of `.getAddress()` method - right now the `PaymentAddress` is returned via an attribute. 
1. The interface on which this attribute is exposed has `.toJSON()`, which itself calls `.toJSON()` on the attribute (returning a nested JSON structure).
1. Less importantly, but something that came up, there are two implementations that have been shipping for about 1 year - so implementers were reluctant to change things in a way that was not backwards compatible.

Reply:
PaymentAddress does seem like a strange mix of dictionaries and classes.

However, as discussed elsewhere I don't think we should be returning dictionaries from APIs in general as it makes it impossible to add methods to them in the future. Using dictionaries as arguments to functions as a way of having optional values sounds fine though.

This might actually be a thing the TAG should have a design principle by itself. @slightlyoff @domenic

Reply:
It doesn't make it impossible; it just means you have to upgrade to a class at that time.